src/algorithm/minkowski/epa/epa3d.rs

```diff
@@ -0,0 +1,357 @@
+/// the support points coming from the A shape.
+fn point<S>(polytope: &Polytope<S>, face: &Face<S>) -> Point3<S>
+where
+    S: BaseFloat,
+{
+    let (u, v, w) = barycentric_vector(
+        face.normal * face.distance,
+        polytope.vertices[face.vertices[0]].v,
+        polytope.vertices[face.vertices[1]].v,
+        polytope.vertices[face.vertices[2]].v,
+    );
+
+    polytope.vertices[face.vertices[0]].sup_a * u
+        + polytope.vertices[face.vertices[1]].sup_a.to_vec() * v
+        + polytope.vertices[face.vertices[2]].sup_a.to_vec() * w
+}
+
+#[derive(Debug)]
+struct Polytope<'a, S: 'a>
+where
+    S: BaseFloat,
+{
+    vertices: &'a mut Vec<SupportPoint<Point3<S>>>,
+    faces: Vec<Face<S>>,
+}
+
+impl<'a, S: 'a> Polytope<'a, S>
+where
+    S: BaseFloat,
+{
+    pub fn new(simplex: &'a mut Vec<SupportPoint<Point3<S>>>) -> Self {
+        let faces = Face::new(simplex);
+        Self {
+            vertices: simplex,
+            faces,
+        }
+    }
+
+    pub fn closest_face_to_origin(&'a self) -> &'a Face<S> {
+        let mut face = &self.faces[0];
+        for f in self.faces[1..].iter() {
+            if f.distance < face.distance {
+                face = f;
+            }
+        }
+        face
+    }
+
+    pub fn add(&mut self, sup: SupportPoint<Point3<S>>) {
+        // remove faces that can see the point
+        let mut edges = Vec::default();
+        let mut i = 0;
+        while i < self.faces.len() {
+            let dot = self.faces[i]
+                .normal
+                .dot(sup.v - self.vertices[self.faces[i].vertices[0]].v);
+            if dot > S::zero() {
+                let face = self.faces.swap_remove(i);
+                remove_or_add_edge(&mut edges, (face.vertices[0], face.vertices[1]));
+                remove_or_add_edge(&mut edges, (face.vertices[1], face.vertices[2]));
+                remove_or_add_edge(&mut edges, (face.vertices[2], face.vertices[0]));
+            } else {
+                i += 1;
+            }
+        }
+
+        // add vertex
+        let n = self.vertices.len();
+        self.vertices.push(sup);
+
+        // add new faces
+        for (a, b) in edges {
+            self.faces.push(Face::new_impl(&self.vertices, n, a, b));
+        }
+    }
+}
+
+#[derive(Debug)]
+struct Face<S> {
+    pub vertices: [usize; 3],
+    pub normal: Vector3<S>,
+    pub distance: S,
+}
+
+impl<S> Face<S>
+where
+    S: BaseFloat,
+{
+    fn new_impl(simplex: &Vec<SupportPoint<Point3<S>>>, a: usize, b: usize, c: usize) -> Self {
+        let ab = simplex[b].v - simplex[a].v;
+        let ac = simplex[c].v - simplex[a].v;
+        let normal = ab.cross(ac).normalize();
+        let distance = normal.dot(simplex[a].v);
+        Self {
+            vertices: [a, b, c],
+            normal,
+            distance,
+        }
+    }
+
+    pub fn new(simplex: &Vec<SupportPoint<Point3<S>>>) -> Vec<Self> {
```

Comment:
same here. I'll stop commenting on that part - please re-evaluate all vector parameters in the PR